Make `Sparse` tensor `Type`s extend `TensorType`

aerubanov opened the pull request:
This pull request is a continuation of the work in PR #303 on refactors the sparse tensor type structure so that sparse types extend the existing tensor types.

Closes #142.

aerubanov commented:
I did a rebase of the branch to the master, which resulted in some of the tests being broken. Now I'm fixing the bugs that occurred, but I decided to open this PR to be able to track progress and discuss changes.

aerubanov commented on aesara/tensor/basic_opt.py:
```diff
@@ -2990,7 +2995,7 @@ def constant_folding(fgraph, node):
         # TODO: `Type` itself should provide an interface for constructing
         # instances appropriate for a given constant.
-        if isinstance(output.type, TensorType):
+        if isinstance(output.type, DenseTensorType):
```
I changed TensorType to DenseTensorType here, because otherwise it trying to create dense constant with sparse data inside.

aerubanov commented on aesara/graph/fg.py:
```diff
@@ -465,7 +465,7 @@ def change_node_input(
         else:
             r = node.inputs[i]
-            if check and not r.type.is_super(new_var.type):
+            if check and not r.type.in_same_class(new_var.type):
```
Previously a method from the Type class was called here that returned true, but now a method from TensorType is called, which now returns false because it additionally checks for a shape match. So I changed the called method.

aerubanov commented:
@brandonwillard , Looks like I fixed tests that were broken after rebase. In your initial PR #303 you noticed that we should "Convert sparse to dense when used with some ```Op```s (e.g. Elemwise, CAReduce)". Could you provide more details about this?

brandonwillard commented:
> @brandonwillard , Looks like I fixed tests that were broken after rebase. In your initial PR #303 you noticed that we should "Convert sparse to dense when used with some `Op`s (e.g. Elemwise, CAReduce)". Could you provide more details about this?

In order to make the old `TensorVariable` interface, `_tensor_py_operators`,  applicable to sparse matrices, we need to start adding support for currently unsupported sparse operations, like `__get__`, which constructs all sorts of `Subtensor` `Op`s.  Right now, sparse matrices implement their own indexing `Op`s, which is a huge interface problem.  We can't reasonably add sparse matrix-specific logic to `Subtensor`, though.

We can get around this by dispatching to the appropriate sparse `Op`s at the `_sparse_py_operators` interface level, but then there are still some related operations that aren't covered, like advanced indexing on sparse matrices.  If we simply convert sparse matrices to dense matrices at that interface, we can&mdash;at the very least&mdash;make things work (i.e. not return graphs that will err when evaluated).

Really, that's all we need to do in order to merge this: make sure the new interface doesn't generate graphs that necessarily produces errors, and converting to a dense matrix is a quick and dirty work-around in most cases.

aerubanov commented:
@brandonwillard ,  I added tests to check that no errors occurred when using ```_tensor_py_operators``` with sparse tensors and added conversion sparse to dense to make them pass. Also, when the conversion happens the warning is generated to inform users about that. Please look at the code and let me know what additional improvements are needed.

brandonwillard reviewed (changes requested):
It looks like you have all the necessary logic in there.  I left a few comments about where certain steps take place (e.g. in the sparse module vs. others).  The basic idea is that we want to keep all sparse object related handling within the sparse module.  In other words, we don't want the concerns of sparse types to leak into code that doesn't explicitly need to concern itself with sparse types.

My guess is that those lines of code can be moved into the sparse `Variable` interface before the external code is called.

Aside from that, we'll need to rebase some of those commits (e.g. combine test-adding commits with their respective code changes, merge/remove commits that undo things performed by earlier commits in this branch, etc.) and update their formatting to match [these guidelines](https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html) (just the summary lines for now).  Since this is such a large set of changes, these kinds of things are even more important&mdash;both for the reviewing overhead and basic organizational reasons.

aerubanov commented:
Hi @brandonwillard ! I made the changes according to your comments. Now I override in ```_sparse_py_operators``` unimplemented operations for sparse variables, converting them to dense. So this logic stays inside the sparse module. And I cleaned up the commit history.

brandonwillard commented:
> Hi @brandonwillard ! I made the changes according to your comments. Now I override in `_sparse_py_operators` unimplemented operations for sparse variables, converting them to dense. So this logic stays inside the sparse module. And I cleaned up the commit history.

Great, thanks!  I'm still a bit backlogged, but I'll review soon.

aerubanov commented:
Hi @brandonwillard! I just wanted to remind you of this PR just in case. Of course, if you have time for that.

brandonwillard commented on aesara/graph/features.py:
```diff
@@ -559,7 +559,7 @@ def replace_all_validate(
         for r, new_r in replacements:
             try:
-                fgraph.replace(r, new_r, reason=reason, verbose=False, **kwargs)
+                fgraph.replace(r, new_r, reason=reason, verbose=verbose, **kwargs)
```
```suggestion
                fgraph.replace(r, new_r, reason=reason, verbose=False, **kwargs)
```

`verbose` is set to `False` here because this method handles that functionality itself.

brandonwillard reviewed (changes requested):
It looks like there were some problems rebasing, because there are a few reversions of upstream changes and unmanaged conflict diffs.

brandonwillard commented:
I'll fix the rebase issues and push in a minute.

brandonwillard reviewed (approved):
There's still more to be done to get the sparse interface into shape, but these changes look like a good step toward that, if only because they explicitly state that we want a more correct type representation.

The biggest issue right now is that `TensorType`s aren't an abstract class, and direct instances of `TensorType`s represent dense tensors, alongside instances of the new `DenseTensor` type.  It might not take much to make such a change and remove all the meta classes, though.

Anyway, I'm fine moving forward with these changes and opening another set of issues to follow up, but let's get a few more eyes on this first.

kc611 commented:
> The biggest issue right now is that TensorTypes aren't an abstract class, and direct instances of TensorTypes represent dense tensors, alongside instances of the new DenseTensor type. It might not take much to make such a change and remove all the meta classes, though.

We should make an issue page for this. Since, this can be tackled separately from the additional logic that we add towards supporting `SparseTypes`.

aerubanov commented:
Hi @brandonwillard! I rebased on top of the master and resolved the conflicts.

brandonwillard commented on tests/test_tensor_interface.py:
```diff
@@ -0,0 +1,159 @@
+        method_to_call = getattr(x, method)
+        z = method_to_call()
+        f = aesara.function([x], z, on_unused_input="ignore")
+        print(f([[1.1, 0.0, 2.0], [-1.0, 0.0, 0.0]]))
```
We require that tests utilize explicit `assert`s.  This helps clarify the expectations of the function being tested and the tests themselves.

brandonwillard commented on tests/test_tensor_interface.py:
```diff
@@ -0,0 +1,159 @@
+import aesara.tensor as at
+
+
+@pytest.mark.parametrize("is_sparse", [False, True])
```
Does the non-sparse case need to be tested explicitly?  It seems like we would necessarily have coverage for the non-sparse/dense cases via the existing `aesara.tensor` tests.

brandonwillard commented:
The location and name of the new `tests.test_tensor_interface` needs to be reconsidered, because it's explicitly testing interfaces that already have corresponding test sub-packages/modules (e.g. `tests.sparse` and `tests.tensor.test_var`).  If the `is_sparse == False` case isn't needed (see [here](https://github.com/aesara-devs/aesara/pull/766#discussion_r827431140)), then this module could be relocated/renamed to `tests.sparse.test_var`.

brandonwillard commented:
I've updated the new tests and added a commit that renames `SparseType` to `SparseTensorType`.

aerubanov commented:
@brandonwillard, Thanks for the review and the corrections!

brandonwillard commented:
@aerubanov, thanks for the PR!  This was an important one that required a lot of persistence.  It's this kind of help that really moves things forward, and we truly appreciate it.
